dali/kernels/signal/moving_mean_square_gpu.cu

```diff
@@ -0,0 +1,376 @@
+// Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cuda_runtime.h>
+#include <cstdio>
+#include <iostream>
+#include <numeric>
+#include <vector>
+#include "dali/core/util.h"
+#include "dali/kernels/signal/moving_mean_square.h"
+#include "dali/kernels/signal/moving_mean_square_gpu.h"
+
+#define SHARED_MEMORY_BANKS 32
+#define LOG_MEM_BANKS 5
+#define CONFLICT_FREE_OFFSET(n) ((n) >> LOG_MEM_BANKS)
+
+namespace dali {
+namespace kernels {
+namespace signal {
+
+namespace {
+
+/**
+ * Type of the accumulator:
+ * In case calculation is performed on floating-point,
+ * we use separate integer to keep the integer part of the number to keep
+ * numeric accuracy.
+ */
+template <typename T>
+struct accumulator_type {
+  using type = int64_t;
+};
+
+template <>
+struct accumulator_type<float> {
+  using type = float;
+};
+
+template <>
+struct accumulator_type<double> {
+  using type = double;
+};
+
+template <typename T>
+using acc_t = typename accumulator_type<T>::type;
+
+}  // namespace
+
+
+struct SampleDesc {
+  void *out;
+  const void *in;
+  int64_t len;
+};
+
+struct conflict_free_pos {
+  DALI_HOST_DEV DALI_FORCEINLINE
+  int operator()(int pos) const noexcept {
+    return pos + CONFLICT_FREE_OFFSET(pos);
+  }
+};
+
+struct square {
+  template <typename T>
+  DALI_HOST_DEV DALI_FORCEINLINE
+  T operator()(T x) const noexcept {
+    return x * x;
+  }
+};
+
+template <typename T>
+struct divide {
+  explicit divide(T divisor) : divisor_(divisor) {}
+
+  DALI_HOST_DEV DALI_FORCEINLINE
+  T operator()(T x) const noexcept {
+    return x / divisor_;
+  }
+
+  T divisor_;
+};
+
+
+template <typename T, typename SharedMemPos = conflict_free_pos>
+__device__ void PrefixSumSharedMem(T* buffer, int pow2,
+                                   SharedMemPos shm_pos = {}) {
+  int offset = 1;
+  int tid = threadIdx.x;
+  for (int d = pow2 >> 1; d > 0; d >>= 1) {  // build sum in place up the tree
```

Comment:
This is built on an example from an ancient book. Now we would rather use warp intrinsics first and only then go to blockwise operations - this would save several `__syncthreads`

Reply:
I agree that this could be improved. I think that we could address such optimizations in a follow-up. Do you agree?